### PR TITLE
[SYCL] Change return of get_native for opencl backend to vector 

### PR DESCRIPTION
**Pull request by denis-kabanov**
Change return of get_native<sycl::backend::opencl>(event) from cl_event to vector<cl_event>

### CODE REVIEW

File: sycl/include/CL/sycl/event.hpp [hunk 1]
```diff
@@ -132,16 +132,32 @@ class __SYCL_EXPORT event {
...
+  detail::enable_if_t<(BackendName == backend::opencl), std::vector<cl_event>>
+  get_native() const {
+    backend_return_t<BackendName, event> ReturnValue;
+    for (auto const &element : getNativeVector()) {
+      ReturnValue.push_back(
+          reinterpret_cast<typename interop<BackendName, event>::value_type>(
+              element));
+    }
+    return ReturnValue;
+  }
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/event.hpp:152]: Why do we need this function at all?

Comment (denis-kabanov) [reply sycl/include/CL/sycl/event.hpp:152]: According to [SYCL 2020 specification (revision 3)](https://www.khronos.org/registry/SYCL/specs/sycl-2020/pdf/sycl-2020.pdf#page=478&zoom=100,64,364) get_native<sycl::backend::opencl>(event) should return vector<cl_event>, whereas for others backends it is not specified. Because C++14 has no "if constexpr", I chose to change SFINAE.

Comment (alexbatashev) [reply sycl/include/CL/sycl/event.hpp:152]: event::get_native is deprecated. We should not change it in any way.

File: sycl/include/CL/sycl/event.hpp [hunk 2]
```diff
@@ -132,17 +132,34 @@ class __SYCL_EXPORT event {
...
+      (BackendName != backend::opencl),
+      typename detail::interop<BackendName, event>::type> get_native() const {
     return reinterpret_cast<typename detail::interop<BackendName, event>::type>(
         getNative());
   }
+  /// \return a native handle vector for opencl according to specification.
+  template <backend BackendName>
+  __SYCL_DEPRECATED("Use SYCL 2020 sycl::get_native free function")
+  detail::enable_if_t<(BackendName == backend::opencl),
+                      std::vector<cl_event>> get_native() const {
```
Comment (cperkinsintel) [inline sycl/include/CL/sycl/event.hpp:145]: shouldn't the type returned by enable_if_t  use `backend_return_t<>` rather than coupling tightly by specifying `std::vector<cl_event>` directly?

Comment (denis-kabanov) [reply sycl/include/CL/sycl/event.hpp:145]: yes, backend_return_t<> is preferred

Comment (denis-kabanov) [reply sycl/include/CL/sycl/event.hpp]: Fixed

File: sycl/include/CL/sycl/event.hpp (hunk 1 above)
Comment (denis-kabanov) [reply sycl/include/CL/sycl/event.hpp]: At this moment, sycl::get_native free function calls a deprecated function with backend=opencl, so to change the free function we need to work with deprecated.

**Review by pvchupin — CHANGES_REQUESTED** [review]
Looks like breaking change. We should hold this until discuss options.

File: sycl/include/CL/sycl/event.hpp (hunk 1 above)
Comment (jbrodman) [reply sycl/include/CL/sycl/event.hpp]: I think Alex is right - I'm not even sure where this public function event::get_native() came from.

Regardless, since it's an existing function with different behavior, we should leave it as-is to avoid the breaking change.

I think we need to change the impl of our free function to not use this deprecated method.

Comment (denis-kabanov) [reply sycl/include/CL/sycl/event.hpp]: @jbrodman is it ok that with any changes API will break, because backend_return_t for opencl event will return std::vector<cl_event> instead of cl_event?

File: sycl/include/CL/sycl/detail/backend_traits_opencl.hpp [hunk 3]
```diff
@@ -109,15 +104,28 @@ template <> struct BackendReturn<backend::opencl, device> {
...
+#else
+template <> struct interop<backend::opencl, event> {
+  using type = std::vector<cl_event>;
+  using value_type = cl_event;
+};
 template <> struct BackendInput<backend::opencl, event> {
   using type = std::vector<cl_event>;
   using value_type = cl_event;
 };
-
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/backend_traits_opencl.hpp:112]: Nit, this is unnecessary change

**Review by alexbatashev — APPROVED** [review]
lgtm

File: sycl/include/CL/sycl/detail/backend_traits_opencl.hpp (hunk 3 above)
Comment (denis-kabanov) [reply sycl/include/CL/sycl/detail/backend_traits_opencl.hpp:112]: Unchanged, BackendReturn template follows BackendInput with the missing line, but now they are in #ifdef block, where all templates (in my realization) go without the missing line. This deletion is a consequence. Would you prefer, missing lines between templates in #ifdef block or not?

**Review by alexbatashev — APPROVED** [review]

**Comment by denis-kabanov** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/682

**Review by alexbatashev — APPROVED** [review]

**Comment by s-kanaev** [comment]
@pvchupin , a friendly ping

**Comment by pvchupin** [comment]
Please update PreprocessorMacros.md

**Comment by denis-kabanov** [comment]
@pvchupin, @intel/llvm-reviewers-runtime, can review this pull request?

**Review by pvchupin — APPROVED** [review]
doc ok

**Review by vladimirlaz — APPROVED** [review]

**Review by alexbatashev — APPROVED** [review]
